src/locator.rs

```diff
@@ -0,0 +1,643 @@
+    const VALID_INPUTS_GIT: &str = r"git\+[^\pC$]+\$[^\pC$]+";
+
+    proptest! {
+        /// Tests randomly generated strings that match the provided regular expression against the parser.
+        /// Validates that the parser succeeds by converting the locator back into a string again.
+        #[test]
+        fn parses_arbitrary_locator_git(input in VALID_INPUTS_GIT) {
+            let parsed = Locator::parse(&input).expect("must parse locator");
+            assert_eq!(parsed.to_string(), input);
+        }
+    }
+
+    /// Regular expression that matches any unicode string that is:
+    /// - Prefixed with `git+`
+    /// - Contains a literal `/`
+    /// - Contains zero or more digits
+    /// - Contains a literal `/`
+    /// - Contains at least one character that is not a control character and not the literal `$`
+    /// - Contains a literal `$`
+    /// - Contains at least one character that is not a control character and not the literal `$`
+    const VALID_INPUTS_GIT_WITH_ORG: &str = r"git\+/\d*/[^\pC$]+\$[^\pC$]+";
+
+    proptest! {
+        /// Tests randomly generated strings that match the provided regular expression against the parser.
+        /// Validates that the parser succeeds by converting the locator back into a string again.
+        #[test]
+        fn parses_arbitrary_locator_git_with_org(input in VALID_INPUTS_GIT_WITH_ORG) {
+            let parsed = Locator::parse(&input).expect("must parse locator");
+            assert_eq!(parsed.to_string(), input);
+        }
+    }
+
+    /// Regular expression that matches any unicode string that is:
+    /// - Prefixed with `custom+`
+    /// - Contains at least one character that is not a control character and not the literal `$`
+    /// - Contains a literal `$`
+    /// - Contains at least one character that is not a control character and not the literal `$`
+    const VALID_INPUTS_CUSTOM: &str = r"custom\+[^\pC$]+\$[^\pC$]+";
+
+    proptest! {
+        /// Tests randomly generated strings that match the provided regular expression against the parser.
+        /// Validates that the parser succeeds by converting the locator back into a string again.
+        #[test]
+        fn parses_arbitrary_locator_custom(input in VALID_INPUTS_CUSTOM) {
+            let parsed = Locator::parse(&input).expect("must parse locator");
+            assert_eq!(parsed.to_string(), input);
+        }
+    }
+
+    /// Regular expression that matches any unicode string that is:
+    /// - Prefixed with `custom+`
+    /// - Contains a literal `/`
+    /// - Contains zero or more digits
+    /// - Contains a literal `/`
+    /// - Contains at least one character that is not a control character and not the literal `$`
+    /// - Contains a literal `$`
+    /// - Contains at least one character that is not a control character and not the literal `$`
+    const VALID_INPUTS_CUSTOM_WITH_ORG: &str = r"custom\+/\d*/[^\pC$]+\$[^\pC$]+";
```

Comment:
Is the `/` after the `+` correct? Custom locators look like `custom+123/foo$1.2.3`